agora/Agora/Proposal/Scripts.hs

```diff
@@ -261,141 +327,130 @@ proposalValidator as maximumCosigners =
-          let sortDatumHashes = phoistAcyclic $ pmsortBy # pltAsData
-
-              sortedStakeInputDatumHashes =
-                sortDatumHashes # stakeInputDatumHashes
-
-              sortedStakeOutputDatumHashes =
-                sortDatumHashes # stakeOutputDatumHashes
-
-          pguardC "All stake datum are unchanged" $
-            plistEquals
-              # sortedStakeInputDatumHashes
-              # sortedStakeOutputDatumHashes
-
-          PPair totalStakedAmount stakeOwners <-
-            pmatchC $
-              pfoldl
-                # plam
-                  ( \l dh -> unTermCont $ do
-                      let stake =
-                            pfromData $
-                              pfromJust
-                                #$ ptryFindDatum @(PAsData PStakeDatum)
-                                # pfromData dh
-                                # txInfoF.datums
-
-                      stakeF <- pletFieldsC @'["stakedAmount", "owner"] $ pto stake
-
-                      PPair amount owners <- pmatchC l
-
-                      let newAmount = amount + punsafeCoerce (pfromData stakeF.stakedAmount)
-                          updatedOwners = pcons # stakeF.owner # owners
-
-                      pure $ pcon $ PPair newAmount updatedOwners
-                  )
-                # pcon (PPair (0 :: Term _ PInteger) (pnil @PBuiltinList))
-                # stakeInputDatumHashes
-
-          sortedStakeOwners <- pletC $ pmsortBy # pltAsData # stakeOwners
-
-          pure $ validationLogic # totalStakedAmount # sortedStakeOwners
-
-    withSingleStake' ::
+        let updateCtx = plam $ \ctx' stake -> unTermCont $ do
+              ctxF <- pmatchC ctx'
+
+              stakeF <-
+                pletFieldsC @'["stakedAmount", "owner"] $
+                  pto stake
+
+              pure $
+                pcon $
+                  PWitneseMultipleStakeContext
+                    { totalAmount =
+                        ctxF.totalAmount
+                          + punsafeCoerce
+                            (pfromData stakeF.stakedAmount)
+                    , orderedOwners =
+                        pcons # stakeF.owner
+                          # ctxF.orderedOwners
+                    }
+
+            f :: Term _ (_ :--> PTxInInfo :--> _)
+            f = plam $ \ctx' ((pfield @"resolved" #) -> txOut) ->
+              pfromMaybe # ctx'
+                #$ (pfmap # (updateCtx # ctx') #$ getStakeDatum # txOut)
+
+            sortOwners = plam $
+              flip pmatch $ \ctxF ->
+                pcon $
+                  PWitneseMultipleStakeContext
+                    { totalAmount = ctxF.totalAmount
+                    , orderedOwners = pmsort # ctxF.orderedOwners
+                    }
+
+            ctx =
+              sortOwners
+                #$ pfoldl
+                # f
+                # pcon (PWitneseMultipleStakeContext 0 pnil)
+                # txInfoF.referenceInputs
+         in plam (# ctx)
```

Comment:
We don't assume that there're only stake inputs in the reference inputs - `getStakeDatum` returns `Nothing` if the given input is not a stake input. But you're right, if there're other stake inputs that have the wrong owners, the checks in `PCosign` and `PAdvanceProposal`, will fail. This is essentially the same before this pr, except that we checked stakes in inputs rather than reference inputs.

Reply:
Okay. I think the issue with other stakes is okay. I'm more concerned about other reference inputs, which seems to be fine.